operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -0,0 +1,134 @@
+			targetReplicas = sset.Replicas(expected)
+		}
+		leaving, removalResult := d.scaleStatefulSetDown(es, actualStatefulSets, &actualStatefulSets[i], targetReplicas, esClient, observedState, reconcileState)
+		results.WithResults(removalResult)
+		if removalResult.HasError() {
+			return results
+		}
+		leavingNodes = append(leavingNodes, leaving...)
+	}
+
+	// migrate data away from nodes leaving the cluster
+	log.V(1).Info("Migrating data away from nodes", "nodes", leavingNodes)
+	if err := migration.MigrateData(esClient, leavingNodes); err != nil {
+		return results.WithError(err)
+	}
+
+	return results
+}
+
+func (d *defaultDriver) scaleStatefulSetDown(
+	es v1alpha1.Elasticsearch,
+	allStatefulSets sset.StatefulSetList,
+	ssetToScaleDown *appsv1.StatefulSet,
+	targetReplicas int32,
+	esClient esclient.Client,
+	observedState observer.State,
+	reconcileState *reconcile.State,
+) ([]string, *reconciler.Results) {
+	results := &reconciler.Results{}
+	logger := log.WithValues("statefulset", k8s.ExtractNamespacedName(ssetToScaleDown))
+
+	if sset.Replicas(*ssetToScaleDown) == 0 && targetReplicas == 0 {
+		// no replicas expected, StatefulSet can be safely deleted
+		logger.Info("Deleting statefulset", "namespace", ssetToScaleDown.Namespace, "name", ssetToScaleDown.Name)
+		if err := d.Client.Delete(ssetToScaleDown); err != nil {
+			return nil, results.WithError(err)
+		}
+	}
+	// copy the current replicas, to be decremented with nodes to remove
+	initialReplicas := sset.Replicas(*ssetToScaleDown)
+	updatedReplicas := initialReplicas
+
+	// leaving nodes names can be built from StatefulSet name and ordinals
+	// nodes are ordered by highest ordinal first
+	var leavingNodes []string
+	for i := initialReplicas - 1; i > targetReplicas-1; i-- {
+		leavingNodes = append(leavingNodes, sset.PodName(ssetToScaleDown.Name, i))
+	}
+
+	// TODO: don't remove last master/last data nodes?
+	// TODO: detect cases where data migration cannot happen since no nodes to host shards?
+
+	for _, node := range leavingNodes {
+		if migration.IsMigratingData(observedState, node, leavingNodes) {
+			// data migration not over yet: schedule a requeue
+			logger.V(1).Info("Data migration not over yet, skipping node deletion", "node", node)
+			results.WithResult(defaultRequeue)
+			// no need to check other nodes since we remove them in order and this one isn't ready anyway
+			break
+		}
+		// data migration over: allow pod to be removed
+		updatedReplicas--
+	}
+
+	if updatedReplicas < initialReplicas {
+		// trigger deletion of nodes whose data migration is over
+		logger.V(1).Info("Scaling replicas down", "from", initialReplicas, "to", updatedReplicas)
+		ssetToScaleDown.Spec.Replicas = &updatedReplicas
+
+		if label.IsMasterNodeSet(*ssetToScaleDown) {
+			// Update Zen1 minimum master nodes API, accounting for the updated downscaled replicas.
+			_, err := zen1.UpdateMinimumMasterNodes(d.Client, es, esClient, allStatefulSets, reconcileState)
```

Comment:
Should this not be done after the downscale for zen1 unless it is the 'dangerous' downscale from 2->1 in which case you _have_ to do it before removing the node?

Reply:
Good point! I think that's what is highlighted by @barkbay in https://github.com/elastic/cloud-on-k8s/issues/668.
So the general idea would be:
1. before deleting a master node: if `mastersCount==2`, update the minimum master nodes (to `1`)
2. always update minimum master nodes at the beginning of nodespecs reconciliation (unless already at the correct value) to account for pods that have been added/removed
3. newly created master nodes start with a configuration that takes into consideration the *expected* number of master nodes (which might be different from the *current* number)

Reply:
I opened a discussion with ES folks, to make sure we get this right: https://github.com/elastic/cloud-on-k8s/issues/1281

Reply:
I'd like to fix this in a next PR if that's OK @pebrc?

Reply:
Sure as long as we track it in an issue!